src/libCrypto/EthCrypto.cpp

```diff
@@ -0,0 +1,183 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "EthCrypto.h"
+
+#include <openssl/ec.h>  // for EC_GROUP_new_by_curve_name, EC_GROUP_free, EC_KEY_new, EC_KEY_set_group, EC_KEY_generate_key, EC_KEY_free
+#include <openssl/obj_mac.h>  // for NID_secp192k1
+#include <openssl/sha.h>      //for SHA512_DIGEST_LENGTH
+#include <ethash/keccak.hpp>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <string>
+
+using namespace std;
+
+// Inspiration from:
+// https://stackoverflow.com/questions/10906524
+// https://stackoverflow.com/questions/57385412/
+// https://medium.com/mycrypto/the-magic-of-digital-signatures-on-ethereum-98fe184dc9c7
+
+// Prefix signed txs in Ethereum with Keccak256("\x19Ethereum Signed
+// Message:\n32" + Keccak256(message))
+constexpr uint8_t prelude[] = {25,  69,  116, 104, 101, 114, 101, 117, 109,
+                               32,  83,  105, 103, 110, 101, 100, 32,  77,
+                               101, 115, 115, 97,  103, 101, 58,  10,  48};
+
+// https://stackoverflow.com/questions/57385412/
+void SetOpensslSignature(const std::string& sSignatureInHex, ECDSA_SIG* pSign) {
+  // Openssl uses raw pointers and macros for freeing but can force it into
+  // smart pointers methodology
+  std::unique_ptr<BIGNUM, std::function<void(BIGNUM*)>> rr(
+      NULL, [](BIGNUM* b) { BN_free(b); });
+  BIGNUM* r_ptr = rr.get();
+  std::unique_ptr<BIGNUM, std::function<void(BIGNUM*)>> ss(
+      NULL, [](BIGNUM* b) { BN_free(b); });
+  BIGNUM* s_ptr = ss.get();
+
+  std::string sSignatureR =
+      sSignatureInHex.substr(0, sSignatureInHex.size() / 2);
+  std::string sSignatureS = sSignatureInHex.substr(sSignatureInHex.size() / 2);
+
+  BN_hex2bn(&r_ptr, sSignatureR.c_str());
+  BN_hex2bn(&s_ptr, sSignatureS.c_str());
+
+  ECDSA_SIG_set0(pSign, r_ptr, s_ptr);
+}
+
+// This function needs to set the key but also needs to decomprss the
+// coordinates using EC_POINT_set_compressed_coordinates_GFp sPubKeyString
+// should be hex \0 terminated string with no compression bit on it
+bool SetOpensslPublicKey(const char* sPubKeyString, EC_KEY* pKey) {
+  // X co-ordinate
+  std::unique_ptr<BIGNUM, std::function<void(BIGNUM*)>> gx(
+      NULL, [](BIGNUM* b) { BN_free(b); });
+
+  BIGNUM* gx_ptr = gx.get();
+
+  EC_KEY_set_asn1_flag(pKey, OPENSSL_EC_NAMED_CURVE);
+
+  // From
+  // https://www.oreilly.com/library/view/mastering-ethereum/9781491971932/ch04.html
+  // The first byte indicates whether the y coordinate is odd or even
+  int y_chooser_bit = 0;
+
+  if (sPubKeyString[1] == '2') {
```

Comment:
where did `sPubKeyString[0]` go?  Why are we starting with `[1]`?

Reply:
That's right - I have added a check for this